relayer/message_coordinator.go

```diff
@@ -0,0 +1,259 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package relayer
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/awm-relayer/database"
+	"github.com/ava-labs/awm-relayer/messages"
+	relayerTypes "github.com/ava-labs/awm-relayer/types"
+	"github.com/ava-labs/subnet-evm/core/types"
+	"github.com/ava-labs/subnet-evm/ethclient"
+	"github.com/ava-labs/subnet-evm/interfaces"
+	"github.com/ava-labs/subnet-evm/precompile/contracts/warp"
+	"github.com/ethereum/go-ethereum/common"
+	"go.uber.org/zap"
+)
+
+// MessageCoordinator contains all the logic required to process messages in the relayer.
+// Other components such as the listeners or the API should pass messages to the MessageCoordinator
+// so that it can parse the message(s) and pass them the the proper ApplicationRelayer.
+type MessageCoordinator struct {
+	logger logging.Logger
+	// Maps Source blockchain ID and protocol address to a Message Handler Factory
+	messageHandlerFactories map[ids.ID]map[common.Address]messages.MessageHandlerFactory
+	applicationRelayers     map[common.Hash]*ApplicationRelayer
+	sourceClients           map[ids.ID]ethclient.Client
+}
+
+func NewMessageCoordinator(
+	logger logging.Logger,
+	messageHandlerFactories map[ids.ID]map[common.Address]messages.MessageHandlerFactory,
+	applicationRelayers map[common.Hash]*ApplicationRelayer,
+	sourceClients map[ids.ID]ethclient.Client,
+) *MessageCoordinator {
+	return &MessageCoordinator{
+		logger:                  logger,
+		messageHandlerFactories: messageHandlerFactories,
+		applicationRelayers:     applicationRelayers,
+		sourceClients:           sourceClients,
+	}
+}
+
+// getAppRelayerMessageHandler returns the ApplicationRelayer that is configured to handle this message, as well as a
+// one-time MessageHandler instance that the ApplicationRelayer uses to relay this specific message.
+// The MessageHandler and ApplicationRelayer are decoupled to support batch workflows in which a single ApplicationRelayer
+// processes multiple messages (using their corresponding MessageHandlers) in a single shot.
+func (mc *MessageCoordinator) getAppRelayerMessageHandler(
+	warpMessageInfo *relayerTypes.WarpMessageInfo,
+) (
+	*ApplicationRelayer,
+	messages.MessageHandler,
+	error,
+) {
+	// Check that the warp message is from a supported message protocol contract address.
+	messageHandlerFactory, supportedMessageProtocol := mc.messageHandlerFactories[warpMessageInfo.UnsignedMessage.SourceChainID][warpMessageInfo.SourceAddress]
```

Comment:
Nit: This line looks very long...I'm not too worried about it specifically but surprised we're not getting a linter error/warning in CI

Reply:
It looks like this can be enabled with the `lll` linter: https://golangci-lint.run/usage/linters/#lll

Reply:
There are a good number of lines that don't fit in the line length, so I'll make a separate ticket for this.